### PR TITLE
feat: Batched Uploads and Green Threads

### PR DESCRIPTION
**Pull request by william-silversmith**
This contains a few features.

1. Send messages to AWS SQS using the send_message_batch (via tq.insert_all) which allows sending 10 messages at a time. This results in a 10x speedup.
2. Add GreenTaskQueue as an option. Users need to manually monkeypatch their application to make it work. This can be useful in NUMA or vCPU environments.
3. Added `delay_seconds` to insert and insert_all which allows you to delay a message's visibility in the queue.

### CODE REVIEW

File: taskqueue/aws_queue_api.py [hunk 1]
```diff
@@ -6,6 +6,17 @@
 
 from .secrets import aws_credentials
 
+
+def toiter(obj):
+  try:
+    iter(obj)
+    if type(obj) is dict:
+      return [ obj ]
```
[the remark this answers is not in the document]
Comment (william-silversmith) [reply taskqueue/aws_queue_api.py]: That's actually what I wanted for this case.

File: taskqueue/aws_queue_api.py [hunk 2]
```diff
@@ -43,13 +54,29 @@ def status(self):
...
-      MessageBody=json.dumps(task),
-    )
-    return resp['MessageId']
+  def insert(self, tasks, delay_seconds=0):
+    tasks = toiter(tasks)
+
+    AWS_BATCH_SIZE = 10 
+
+    resps = []
+    for i in range(0, len(tasks), AWS_BATCH_SIZE):
```
[the remark this answers is not in the document]
Comment (william-silversmith) [reply taskqueue/aws_queue_api.py]: This is true and a headache. I should call it "tolist".

File: taskqueue/aws_queue_api.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (nkemnitz) [reply taskqueue/aws_queue_api.py]: Got it. If it's not a "general purpose" function, you could try to hide it a bit from the user - maybe name it `_to_iterable`. Not crucial, though.

File: taskqueue/taskqueue.py [hunk 3]
```diff
@@ -320,6 +253,246 @@ def block_until_empty(self, interval_sec=2):
...
+      region, qurl, n_threads, project
+    )
+    self._pool = gevent.pool.Pool(n_threads)
+    self.check_monkey_patch_status()
+
+
+  def check_monkey_patch_status(self):
+    import gevent.monkey
+    if not gevent.monkey.is_module_patched("socket"):
+      print(yellow("""
```
[the remark this answers is not in the document]
Comment (william-silversmith) [reply taskqueue/taskqueue.py]: fixed.

**Comment by william-silversmith** [comment]
@nkemnitz latest changes include allowing the use of generators instead of lists in `insert_all` for both GreenTaskQueue and TaskQueue. Additionally, we do not monkey patch automatically. Instead, if the `socket` standard library is not monkey patched and you are using GreenTaskQueue, we issue a warning containing instructions on how to monkey patch your application. TaskQueue continues to work.

File: taskqueue/aws_queue_api.py [hunk 4]
```diff
@@ -6,6 +6,17 @@
 
 from .secrets import aws_credentials
 
+
+def toiter(obj):
```
[the remark this answers is not in the document]
Comment (william-silversmith) [reply taskqueue/aws_queue_api.py]: Now I am!

File: test/test_taskqueue.py [hunk 5]
```diff
@@ -6,7 +6,7 @@
 import pytest
 
 import taskqueue
-from taskqueue import RegisteredTask, TaskQueue, MockTask, PrintTask, LocalTaskQueue, MockTaskQueue
+from taskqueue import RegisteredTask, GreenTaskQueue, TaskQueue, MockTask, PrintTask, LocalTaskQueue, MockTaskQueue
```
Comment (william-silversmith) [inline test/test_taskqueue.py]: No monkey patching here, so GreenTaskQueue will simply behave as a single threaded TaskQueue. monkey patching was disrupting LocalTaskQueue.

**Pull request merged**